File: research/ntrip_ingestion/src/nlgcp_ntrip_ingest/framing.py
"""Streaming RTCM 3.x framer for live TCP chunks.

Reuses the tested Phase 9 CRC-24Q and message-number conventions; the
difference is incrementality: arbitrary TCP splits are buffered until a
full frame is available, and multiple frames per read are all emitted.
Never loads a whole live stream into memory (O(max frame) buffering).
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass

from nlgcp_rtcm_replay.crc import crc24q
from nlgcp_rtcm_replay.framing import (
    CRC_LENGTH,
    HEADER_LENGTH,
    MAX_RTCM_LENGTH,
    PREAMBLE,
    extract_message_number,
)


@dataclass(slots=True)
class StreamedFrame:
    """One validated transport frame with its global byte offset."""

    byte_offset: int
    raw: bytes
    message_number: int | None
    crc_ok: bool
    frame_sha256: str


class StreamingFramer:
    """Incremental RTCM 3.x framer over arbitrary TCP chunk boundaries."""

    def __init__(self, *, max_frame_length: int = MAX_RTCM_LENGTH) -> None:
        if not 1 <= max_frame_length <= MAX_RTCM_LENGTH:
            raise ValueError(f"max_frame_length must be within 1..1023: {max_frame_length}")
        self._max_frame_length = max_frame_length
        self._buffer = bytearray()
        self._byte_offset = 0
        self.findings: list[str] = []
        self.stray_bytes = 0
        self._pending_stray = 0

    @property
    def pending_bytes(self) -> int:
        return len(self._buffer)

    def feed(self, chunk: bytes) -> list[StreamedFrame]:
        """Consume one TCP read; return zero or more complete frames."""
        self._buffer.extend(chunk)
        frames: list[StreamedFrame] = []
        skipped = self._pending_stray
        self._pending_stray = 0
        while True:
            if not self._buffer:
                break
            if self._buffer[0] != PREAMBLE:
                skipped += 1
                self._buffer.pop(0)
                self._byte_offset += 1
                continue
            if len(self._buffer) < HEADER_LENGTH:
                break
            length = ((self._buffer[1] & 0x03) << 8) | self._buffer[2]
            if length > self._max_frame_length:
                self.findings.append(
                    f"OVERSIZED_LENGTH offset={self._byte_offset} length={length}"
                )
                skipped += 1
                self._buffer.pop(0)
                self._byte_offset += 1
                continue
            need = HEADER_LENGTH + length + CRC_LENGTH
            if len(self._buffer) < need:
                self._pending_stray = skipped
                break
            raw = bytes(self._buffer[:need])
            body, crc_bytes = raw[:-CRC_LENGTH], raw[-CRC_LENGTH:]
            expected = (crc_bytes[0] << 16) | (crc_bytes[1] << 8) | crc_bytes[2]
            crc_ok = crc24q(body) == expected
            payload = raw[HEADER_LENGTH:-CRC_LENGTH] if length else b""
            message_number = extract_message_number(payload)
            if skipped:
                self.stray_bytes += skipped
                self.findings.append(
                    f"STRAY_BYTES skipped={skipped} next_offset={self._byte_offset}"
                )
                skipped = 0
            frames.append(
                StreamedFrame(
                    byte_offset=self._byte_offset,
                    raw=raw,
                    message_number=message_number,
                    crc_ok=crc_ok,
                    frame_sha256=hashlib.sha256(raw).hexdigest(),
                )
            )
            if not crc_ok:
                self.findings.append(
                    f"CRC_FAIL offset={self._byte_offset} length={length} "
                    f"message={message_number}"
                )
            del self._buffer[:need]
            self._byte_offset += need
        if skipped:
            # Non-preamble bytes already consumed from the wire that never
            # resynchronised (EOF or waiting for more data) are reported
            # as stray so no input byte is silently ignored.
            self.stray_bytes += skipped
            self.findings.append(
                f"STRAY_BYTES skipped={skipped} next_offset={self._byte_offset + len(self._buffer)}"
            )
        return frames

    def flush_findings(self) -> list[str]:
        findings = list(self.findings)
        self.findings.clear()
        return findings

File: services/gnss-ingestor/internal/ntrip/framing.go
package ntrip

// RTCM 3.x transport constants: 0xD3 preamble, 2 header bytes carrying
// 6 reserved bits + 10-bit payload length, 3-byte CRC-24Q trailer.
const (
	Preamble     = 0xD3
	HeaderLength = 3
	CRCLength    = 3
	MaxRTCMLen   = 1023
)

// Frame is one validated transport frame with its global byte offset.
type Frame struct {
	ByteOffset    int64
	Raw           []byte
	MessageNumber *int
	CRCOK         bool
	SHA256        [32]byte
}

// Framer incrementally parses RTCM 3.x over arbitrary TCP chunk
// boundaries: split frames are buffered, multiple frames per read are
// all emitted, and no whole live stream is ever held in memory.
type Framer struct {
	maxFrameLength int
	buffer         []byte
	byteOffset     int64
	pendingStray   int
	Findings       []string
	StrayBytes     int
}

// NewFramer returns a streaming framer bounded by maxFrameLength (1..1023).
func NewFramer(maxFrameLength int) (*Framer, error) {
	if maxFrameLength < 1 || maxFrameLength > MaxRTCMLen {
		return nil, &ConfigError{Field: "max_frame_length"}
	}
	return &Framer{maxFrameLength: maxFrameLength}, nil
}

// PendingBytes reports bytes buffered awaiting a complete frame.
func (f *Framer) PendingBytes() int { return len(f.buffer) }

// Feed consumes one TCP read and returns zero or more complete frames.
func (f *Framer) Feed(chunk []byte) []Frame {
	f.buffer = append(f.buffer, chunk...)
	var frames []Frame
	skipped := f.pendingStray
	f.pendingStray = 0
	for {
		if len(f.buffer) == 0 {
			break
		}
		if f.buffer[0] != Preamble {
			skipped++
			f.buffer = f.buffer[1:]
			f.byteOffset++
			continue
		}
		if len(f.buffer) < HeaderLength {
			break
		}
		length := int(f.buffer[1]&0x03)<<8 | int(f.buffer[2])
		if length > f.maxFrameLength {
			f.Findings = append(f.Findings, itoa("OVERSIZED_LENGTH offset=", f.byteOffset, length))
			skipped++
			f.buffer = f.buffer[1:]
			f.byteOffset++
			continue
		}
		need := HeaderLength + length + CRCLength
		if len(f.buffer) < need {
			f.pendingStray = skipped
			break
		}
		raw := append([]byte(nil), f.buffer[:need]...)
		body, crcBytes := raw[:need-CRCLength], raw[need-CRCLength:]
		crcOK := VerifyFrameCRC(body, crcBytes)
		var payload []byte
		if length > 0 {
			payload = raw[HeaderLength : need-CRCLength]
		}
		msg := ExtractMessageNumber(payload)
		if skipped > 0 {
			f.StrayBytes += skipped
			f.Findings = append(f.Findings, itoa("STRAY_BYTES skipped=", int64(skipped), 0)+itoa(" next_offset=", f.byteOffset, 0))
			skipped = 0
		}
		frames = append(frames, Frame{
			ByteOffset:    f.byteOffset,
			Raw:           raw,
			MessageNumber: msg,
			CRCOK:         crcOK,
			SHA256:        SHA256Sum(raw),
		})
		if !crcOK {
			f.Findings = append(f.Findings, itoa("CRC_FAIL offset=", f.byteOffset, length))
		}
		f.buffer = f.buffer[need:]
		f.byteOffset += int64(need)
	}
	if skipped > 0 {
		f.StrayBytes += skipped
		f.Findings = append(f.Findings, itoa("STRAY_BYTES skipped=", int64(skipped), 0))
	}
	return frames
}

// ExtractMessageNumber returns the 12-bit RTCM message number where
// technically valid (payload of at least 2 bytes), else nil.
func ExtractMessageNumber(payload []byte) *int {
	if len(payload) < 2 {
		return nil
	}
	msg := (int(payload[0]) << 4) | ((int(payload[1]) >> 4) & 0xFFF)
	return &msg
}
